ipa-core/src/net/test.rs

```diff
@@ -9,46 +9,171 @@
 
 #![allow(clippy::missing_panics_doc)]
 use std::{
-    array,
+    collections::HashSet,
     net::{SocketAddr, TcpListener},
+    ops::Index,
 };
 
 use once_cell::sync::Lazy;
 use rustls_pki_types::CertificateDer;
 
-use super::transport::MpcHttpTransport;
+use super::{transport::MpcHttpTransport, ConnectionFlavor, Shard};
 use crate::{
     config::{
         ClientConfig, HpkeClientConfig, HpkeServerConfig, NetworkConfig, PeerConfig, ServerConfig,
         TlsConfig,
     },
     executor::{IpaJoinHandle, IpaRuntime},
-    helpers::{HandlerBox, HelperIdentity, RequestHandler},
+    helpers::{HandlerBox, HelperIdentity, RequestHandler, TransportIdentity},
     hpke::{Deserializable as _, IpaPublicKey},
     net::{ClientIdentity, Helper, MpcHelperClient, MpcHelperServer},
+    sharding::{ShardIndex, ShardedHelperIdentity},
     sync::Arc,
     test_fixture::metrics::MetricsHandle,
 };
 
-pub const DEFAULT_TEST_PORTS: [u16; 3] = [3000, 3001, 3002];
+/// Simple struct to keep default port configuration organized.
+#[derive(Clone)]
+pub struct Ports {
+    ring: [u16; 3],
+    shards: [u16; 3],
+}
 
-pub struct TestConfig {
-    pub disable_https: bool,
-    pub network: NetworkConfig<Helper>,
-    pub servers: [ServerConfig; 3],
-    pub sockets: Option<[TcpListener; 3]>,
+/// A **single** ring with 3 hosts, each with a ring and sharding port.
+pub const DEFAULT_TEST_PORTS: Ports = Ports {
+    ring: [3000, 3001, 3002],
+    shards: [6000, 6001, 6002],
+};
+
+/// Configuration of a server that can be reached via socket or port.
+pub struct AddressableTestServer {
+    pub id: ShardedHelperIdentity,
+    /// Contains the ports
+    pub config: ServerConfig,
+    /// Sockets are created if no port was specified.
+    pub socket: Option<TcpListener>,
```

Comment:
as much as I like the idea of holding onto TcpListener until the moment when server is spawned - minimizes the chances of Kernel re-using that port in between - I don't think the approach of storing `TcpListener` in the configuration struct is appropriate here. `TcpListener` does not describe the configuration, it actively uses a system resource (socket)

Maybe we can get away with using file descriptors, but honestly I wouldn't bother as I've never seen an issue with bind/release.

Reply:
Ahh, I think that might be problematic with many tests running at the same time.

Reply:
Kernel is good not releasing the port right away